File: JavaSolutions/P09/ValidBinaryTree.java
/* Write a function to check that a binary tree ↴ is a valid binary search tree */

import java.util.*;

public class ValidBinaryTree {

    public static class BinaryTreeNode {
	
	public int value;
	public BinaryTreeNode left;
	public BinaryTreeNode right;
	
	public BinaryTreeNode(int value) {
	    this.value = value;
	}
	
	public BinaryTreeNode insertLeft(int leftValue) {
	    this.left = new BinaryTreeNode(leftValue);
	    return this.left;
	}
	
	public BinaryTreeNode insertRight(int rightValue) {
	    this.right = new BinaryTreeNode(rightValue);
	    return this.right;
	}
    }

    public static void dft(BinaryTreeNode n, ArrayList<Integer> nodes) {
	if (n == null) return;
	dft(n.left, nodes);
	nodes.add(n.value);
	dft(n.right, nodes);
    }
    
    public static boolean isValid(BinaryTreeNode n) {
	if (n == null) return true;
	ArrayList<Integer> nodes = new ArrayList<Integer>();
	dft(n, nodes);
	System.out.println("DFT: " + Arrays.toString(nodes.toArray()));
	int prev = nodes.get(0);
	for (int i = 1; i < nodes.size(); i++) {
	    if (nodes.get(i) < prev) return false;
	    prev = nodes.get(i);
	}
	return true;
    }

    static int prev = Integer.MIN_VALUE;
    static boolean valid = true;
    public static void isValidEff(BinaryTreeNode n) {
	if (n == null) return;
	isValidEff(n.left);
	if (n.value < prev) {
	    valid = false;
	    return;
	}
	prev = n.value;
        isValidEff(n.right);
    }

    public static void main(String[] args) {
	BinaryTreeNode n7 = new BinaryTreeNode(7);
	BinaryTreeNode n3 = n7.insertLeft(3);
	BinaryTreeNode n10 = n7.insertRight(10);
	BinaryTreeNode n2 = n3.insertLeft(2);
	BinaryTreeNode n6 = n3.insertRight(6);
	BinaryTreeNode n1 = n2.insertLeft(1);
	BinaryTreeNode n5 = n6.insertLeft(5);
	BinaryTreeNode n4 = n5.insertLeft(4);
	BinaryTreeNode n8 = n10.insertLeft(8);
	BinaryTreeNode n11 = n10.insertRight(11);
	BinaryTreeNode n9 = n8.insertRight(9);

	System.out.println(isValid(n7) ? "Valid" : "Invalid");

	isValidEff(n7);
	System.out.println(valid ? "Valid" : "Invalid");
    }
}

File: PythonSolutions/P09/ValidBinaryTree.py
# Write a function to check that a binary tree
# is a valid binary search tree

class BinaryTreeNode:
	def __init__(self, value, left = None, right = None):
		self.value = value
		self.left = left
		self.right = right

prev = None
def isValidBST(root):
	global prev
	if root is None: return True
	if not isValidBST(root.left): return False 
	if prev != None and root.value <= prev.value: return False 
	prev = root
	if not isValidBST(root.right): return False
	return True

def test(root):
	result = "is" if isValidBST(root) else "is not"
	print "n%d %s a valid BST" % (root.value, result)

# Test for the following tree:
#          6
#        /   \
#       /     \
#      4       8
#     / \     /
#    2   5   7
#   / \
#  1   3
n1 = BinaryTreeNode(1)
n3 = BinaryTreeNode(3)
n2 = BinaryTreeNode(2, n1, n3)
n5 = BinaryTreeNode(5)
n4 = BinaryTreeNode(4, n2, n5)
n7 = BinaryTreeNode(7)
n8 = BinaryTreeNode(8, n7, None)
n6 = BinaryTreeNode(6, n4, n8)
test(n6)